### PR TITLE
MM-21925: Enforces 'create_post' permission.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary

Enforces the `create_post` permission in the mobile app.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-21925

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes

#### Device Information

iPhone 11 Simulator running iOS 13.3

#### Screenshots

Simulator Screen Shot - iPhone 11 - 2020-02-10 at 07 11 20
Simulator Screen Shot - iPhone 11 - 2020-02-10 at 07 11 13

### CODE REVIEW

File: app/components/post/index.js [hunk 1]
```diff
@@ -52,6 +56,15 @@ function makeMapStateToProps() {
         let isFirstReply = true;
         let isLastReply = true;
         let commentedOnPost = null;
+        const channel = getChannel(state, {id: post.channel_id}) || {};
+        const canPost = haveIChannelPermission(
+            state,
+            {
+                channel: post.channel_id,
+                team: channel.team_id,
+                permission: Permissions.CREATE_POST,
```
Comment (enahum) [inline app/components/post/index.js]: what server version was this added? do we need to check for the server version before checking for the permission?

Comment (mkraft) [reply app/components/post/index.js]: This is not a new permission so I assumed we don't need to check. It was part of the original set of permissions released about 2 years ago, but was never enforced on the web and mobile clients.

**Comment by chuttam** [comment]
It looks like `mattermost-redux` has [post_utils](https://github.com/mattermost/mattermost-redux/blob/master/src/utils/post_utils.ts) with `canEditPost` and `canDeletePost` functions to check `haveIChannelPermission`. 

a) Is it  feasible to have this create posts permission check follow the pre-existing pattern like edit and delete?

b) I went down this thought path during PR review because I wondered what would happen if a user tried to edit a post in a channel where he/she doesn't have `CREATE_POST` as a permission. Should `(!canPost)` automatically mean `canEdit = false`, just as how you have `canReply` set to `false` in this case?

**Comment by mkraft** [comment]
> It looks like `mattermost-redux` has [post_utils](https://github.com/mattermost/mattermost-redux/blob/master/src/utils/post_utils.ts) with `canEditPost` and `canDeletePost` functions to check `haveIChannelPermission`.
> 
> a) Is it feasible to have this create posts permission check follow the pre-existing pattern like edit and delete?
> 
> b) I went down this thought path during PR review because I wondered what would happen if a user tried to edit a post in a channel where he/she doesn't have `CREATE_POST` as a permission. Should `(!canPost)` automatically mean `canEdit = false`, just as how you have `canReply` set to `false` in this case?

`canEditPost` and `canDeletePost` seem to be evaluating additional business logic and optimizations. For example, if you're the post owner the code path is slightly different from if you're not. Also system messages have different rules. Conversely the `create_post` permission check has no pre-existing post to operate against for optimizations and has not "is owner" code paths, so for those reasons I don't see a helper being necessary.

In terms of making the permissions dependent, the permissions are not currently dependent upon each other on the server-side and I think that would be the correct place to create that dependency. For now if they want to, say, allow creating post and then remove that permissions but allow users to edit their old posts: they can. Worth discussing though, especially when we go and expose the `create_post` in the system console.

**Comment by chuttam** [comment]
a) I'm sold on "create post" not having any setup logic to warrant a helper.

b) Agree, keeping the create and edit posts permissions independent makes sense, with granular control.

File: app/components/post/post.js [hunk 2]
```diff
@@ -263,7 +266,7 @@ export default class Post extends PureComponent {
         const style = getStyleSheet(theme);
         const isReplyPost = this.isReplyPost();
         const onUsernamePress =
-            Config.ExperimentalUsernamePressIsMention && !channelIsReadOnly ? this.autofillUserMention : this.viewUserProfile;
+            (Config.ExperimentalUsernamePressIsMention && !channelIsReadOnly && canPost) ? this.autofillUserMention : this.viewUserProfile;
```
Comment (enahum) [inline app/components/post/post.js]: I'm inclined to remove `ExperimentalUsernamePressIsMention` this was added by shift and is experimental and only applies on build time. If we remove it we wouldn't need the  `canPost` or `channelIsReadOnly` props.

@esethna thoughts?

**Comment by srkgupta** [comment]
@mkraft can you please fix the conflicting files

File: app/components/post/post.js (hunk 2 above)
Comment (mkraft) [reply app/components/post/post.js]: @enahum @esethna Just wondering if it would be OK to decouple the above decision from this feature/PR?

Comment (enahum) [reply app/components/post/post.js]: I'm 100% on this, but lets wait for @esethna input

Comment (esethna) [reply app/components/post/post.js]: Agree on not blocking this PR on this decision. 

I'm not sure what "ExperimentalUsernamePressIsMention" is, but if it's an experimental config setting we can't just drop support for it without warning customers and pre-communicating

Comment (enahum) [reply app/components/post/post.js]: @esethna no customer other than shift uses this AND you know what happened with shift, right?

won;t block this PR but I'm submitting another PR to get rid of this right now.

Comment (enahum) [reply app/components/post/post.js]: After some discussion with @esethna I'll be submitting a PR to get rid of the feature. @mkraft you can revert the changes on this file for `canPost`

File: app/components/post_body/post_body.js [hunk 3]
```diff
@@ -71,6 +71,7 @@ export default class PostBody extends PureComponent {
         shouldRenderJumboEmoji: PropTypes.bool.isRequired,
         theme: PropTypes.object,
         location: PropTypes.string,
+        canPost: PropTypes.bool.isRequired,
```
Comment (enahum) [inline app/components/post_body/post_body.js]: do you need this here?

Comment (mkraft) [reply app/components/post_body/post_body.js]: No, the only place where `<PostBody>` is used doesn't pass it in as a prop, and anyway the `PostOptions` screen looks it up. I can either: 

1) Remove this prop and don't pass it to to the `PostOptions` screen, or
2) Change it to make post_body/index.js lookup the `canPost` value in Redux, leave this prop here, and pass it all the way down to the `PostOptions` screen and remove the lookup there.

Preference?

Comment (enahum) [reply app/components/post_body/post_body.js]: @mkraft I rather have it looked up by the post_options connector as it only happens one time and when needed and that way we don't need to update every single postBody when this value changes

Comment (enahum) [reply app/components/post_body/post_body.js]: @mkraft let me done when this is removed so I can give it another look.

Comment (mkraft) [reply app/components/post_body/post_body.js]: It's pushed.

File: app/components/post_body/post_body.test.js [hunk 4]
```diff
@@ -46,6 +46,7 @@ describe('PostBody', () => {
         isEmojiOnly: false,
         shouldRenderJumboEmoji: false,
         theme: Preferences.THEMES.default,
+        canPost: true,
```
Comment (enahum) [inline app/components/post_body/post_body.test.js]: I think you forgot to remove this one here

**Comment by mkraft** [comment]
@matthewbirtch This ticket is not intended to fix existing (non-regression) styling bugs. Please file a ticket for triage, which will likely be assigned to SET or the mobile team for dev and QA. Thanks.

**Comment by matthewbirtch** [comment]
Ok, I was assigned to review the PR, but I'm not sure what I'm supposed to review then?

**Comment by enahum** [comment]
@esethna any reason why you added a UX review for a PR that is not related to UX or was that a mistake?

**Comment by esethna** [comment]
@enahum ah, I thought there was new styling added. But please file a ticket as an improvement @matthewbirtch

**Comment by enahum** [comment]
@srkgupta the disabled UI would be another ticket and it should go to UX first.

As for the paste action, have you tried on an actual device? if using a simulator does it repro on current master?

the paste functionality is not reproducing for me on latest 1.29 build 276 on the actual device.

**Comment by srkgupta** [comment]
Hi @enahum 
Please note that the above mentioned issues are specific to this PR. The PR is supposed to handle this and hence these changes would not be available on the current master. 

Also I am not sure why we would need a seperate PR just to handle the disabled UI, as this PR is supposed to disable the post input box if the Channel is marked as read-only in Channel Moderation Widget Settings (new feature for 5.22)

**Comment by enahum** [comment]
@srkgupta there is already support for read only channels that were added as part of the Experimental feature for Town-Square a while ago, and the UI has been the same since, in any case the ticket does not mention anything about UI it wasn't looked at by UX to define the UI changes and the code definitely does not have any UI changes (not sure why the description says it has).

Also the paste functionality on the read-only field is already there in current master as well, that is why I brought it up.

**Comment by srkgupta** [comment]
Thanks for the explanation @enahum. I got it now. As you had mentioned, when Town-Square is read-only, I am unable to paste any content on the post input on a real device. This issue is only seen on simulator. So we can ignore this. 

With regards to the greyed out for disabled input, since we are reusing the existing component in this PR, the issue will be separate from this PR. I will raise it as a separate issue and let UX handle this.

**Comment by srkgupta** [comment]
The following issue has been raised separately:

RN: When a Channel is Read-Only, the Post Input should be greyed out like in webapp
https://mattermost.atlassian.net/browse/MM-23224

**Pull request merged**